### ISSUE TITLE
User Accounts for AdGuardHome

### ISSUE THREAD

**Issue by ArchiveBase** [AdguardTeam/AdGuardHome#997]
Add multi-user system that handles multiple simultaneous active sessions with role of administrators and standard users.

Benefits
- Configure non-privilege users (view mode) to access query logs, dashboard and settings.
- Provide non-privilege access to certain device or range.
- Configure administrator access to certain 'Client' https://github.com/AdguardTeam/AdGuardHome/issues/809 then able manage data and change settings (change DNS https://github.com/AdguardTeam/AdGuardHome/issues/821 and filters https://github.com/AdguardTeam/AdGuardHome/issues/435 etc)

![Screenshot_247](https://user-images.githubusercontent.com/46082456/64710761-b839bd80-d4e2-11e9-85fc-427c68adf2cf.png)
_(Source: Ntopng)_

https://www.ntop.org/products/traffic-analysis/ntop/
https://github.com/ntop/ntopng

_Any other reasons?
 To prevent implement https://github.com/AdguardTeam/AdGuardHome/issues/628_

**Comment by szolin** [comment]
First we need to decide whether we can use the same UI for both administrator and a regular user.  The problem is that currently if the server starts to respond with an error to all requests except `get /querylog` UI will show tons of error messages.  So either:
* UI must know what is allowed to a regular user and make only allowed requests
* or UI must handle access errors to any request and ignore it silently (and show empty data everywhere)
* or UI needs a whole new interface for a regular user

**Comment by ameshkov** [comment]
@szolin re-assigned this task to v0.103, I think it's too early for us to go this deep with users management.

**Comment by AngelFalse** [comment]
I didn't quite get the point of adding the "Language" settings into this panel. Based on my experience on the current version, obtaining language from the browser's request is quite enough.

**Comment by ArchiveBase** [comment]
@szolin
> * UI must know what is allowed to a regular user and make only allowed requests
> * or UI needs a whole new interface for a regular user

Prefer explain few more things,
-  The person installed adguardhome must be the 'super administrator'. It means first account only has permission to delete or manage all administrators. I think it's not necessary to provide this special privilege for multiple accounts. simply he/she can share the password because it doesn't make any sense about safety or privacy
-  Let normal administrator to create new administrators, standard users and delete them but it's only if allowed (see https://github.com/AdguardTeam/AdGuardHome/issues/1245) and for assigned ranges or networks by 'super administrator' or 'inherited administrator'
_(these administrators cannot remove or change permissions of other administrators)_


**Clarify above benefits due misunderstand in https://github.com/AdguardTeam/AdGuardHome/issues/1235** 
(basically each user is unique with specific access control and any changes made by certain user only affect to anything assigned not globally)

> * Configure non-privilege users (view mode) to access query logs, dashboard and settings.

Such non-privilege user (standard user) able to see everything same as super administrator but cannot change any settings. (it's easy to configure similar above screenshot)

> * Provide non-privilege access to certain device or range.

It means configure view mode for certain IP or range. (super administrator, administrators and standard users must have ability to create that type account (see more details https://github.com/AdguardTeam/AdGuardHome/issues/1245)

> * Configure administrator access to certain 'IPs, CIDR, MACs' #809 for manage data and change settings (change DNS #821, filters #435 etc)
(this way we can keep provide 'Client' feature for such administrator)

Just super administrator and administrators capable to create that type accounts. 
_(It fulfill https://github.com/AdguardTeam/AdGuardHome/issues/1235 means if one of adguardhome server administrator created separate administrator account with access related to my computer and phone then i can use it for see query log, add or remove filter rules, change settings)_

---

**I forgot something to mention then we need second step for this account create wizard otherwise assigned devices loss protection till new administrator login and configure it**
-  For add necessary filters
-  For add necessary DNS servers
-  For configure general settings
-  For configure blocked services

---

I extend these features (third step) to be more user friendly in separate issue because @ameshkov may worry that's already too complex 😉

**Comment by mountainsandcode** [comment]
Like this idea - as a second iteration of this it could be nice to be able to have users authenticate through some SSO or LDAP service rather than having to manage them manually

**Comment by ArchiveBase** [comment]
> @szolin 
First we need to decide whether we can use the same UI for both administrator and a regular user.

Please refer new changes above; i forgot to mention

**Comment by timcappalli** [comment]
> Like this idea - as a second iteration of this it could be nice to be able to have users authenticate through some SSO or LDAP service rather than having to manage them manually

Agreed. OAuth 2.0 / OpenID Connect and SAML support would be amazing. There's a ton of awesome ready to go libraries out there.

**Comment by ainar-g** [comment]
From #3243:

>    1. Add, remove and disable users;
> 
>    2. Password changes and disable it;
> 
>    3. IP and subnet restrictions (incl. access only from local networks or via unique url maybe, idk);
> 
>    4. Session frame time options (web_session_ttl?);
> 
>    5. Disable authorization entirely?

**Comment by ainar-g** [comment]
From #3806:

 >  I would like the ability to create additional read-only users.
 >  I would like the ability to grant access to the */control/querylog path

**Comment by fernvenue** [comment]
From https://github.com/AdguardTeam/AdGuardHome/issues/4426:

> Can you provide a way (in the yaml file) to actually disable the native authentication process for these kind of settings.

**Comment by b255ea006** [comment]
I think that this feature a multi user setup would be an overshoot and is not worth developing. 

Regards

**Comment by FallenAngel666** [comment]
I would love this feature. Right now if I want to fetch AdGuard data like in Heimdal or other integrations, I need to provide my username and paassword which is used for admin access. A read only user would be much more secure. 

I guess this must not be an overshoot or overengineered solution with different ui or anything. Just a check if the user is admin or not on any set/update operations.

**Comment by mietzen** [comment]
> I would love this feature. Right now if I want to fetch AdGuard data like in Heimdal or other integrations, I need to provide my username and paassword which is used for admin access. A read only user would be much more secure.
> 
> I guess this must not be an overshoot or overengineered solution with different ui or anything. Just a check if the user is admin or not on any set/update operations.

Same here, I think a simple auth token would be sufficient. Ideally the access rights of the token could be restricted to e.g. only show info. I also think implementing a whole multi user system would be a massive overshoot.

**Comment by joeyfigaro** [comment]
I'd love to have this! It'd be great if I could grant my partner access to our AGH instance to help manage our kid's stuff.

**Comment by HVR88** [comment]
It's only been 4 years, so I guess this is still in the consideration stage?

root user default (and only) login is not within acceptable use policy for any such service.

**Comment by lusky3** [comment]
Not hopeful, seeing the dates, but also going to follow this. My partner needs the ability to whitelist to fix broken sites when I'm not available, and I don't need the worry of her breaking something unintentionally.

**Comment by feelfreelinux** [comment]
Not sure if it was mentioned before, but I was able to work this around by adding an another user and a bcrypt-hashed password to `AdGuardHome.yaml`

as in:
```yaml
...
users:
 - name: user1
     password: $2a$12$fnLMhmQKBmybNeAkvBiO3O1KOtXk.JCMFZlPHd5q5AveYCkPCNbg.
 - name: user2
     password: $2a$12$ySnR0uOO.PZdCy5lii0BH.2Znye2Ulsgosb5A7vUqbTcTkjpg2b5a

```

with this I can sign in with two different credentials to AGH's dashboard

**Comment by cquike** [comment]
> Not sure if it was mentioned before, but I was able to work this around by adding an another user and a bcrypt-hashed password to `AdGuardHome.yaml`

That's at least something. But you are not yet able to have one of the users as read-only, for instance, are you?
Thanks for the tip anyway!

**Comment by bwp85** [comment]
Most would think it not necessary to provide this level of configuration. For one I think it would be a great option for below cases.

- User level account for under aged household members so they cannot modify restriction filters, however they could have access to allow Ad based sites etc.
- Restricted admin user for parental controls
- Separate (Super) admin account for remote management / support

Adguard home is generally set and forget so I can see why development has not been focused on a multiuser configuration.
